### PR TITLE
Added serialization/deserialization functions

### PR DESCRIPTION
**Pull request by lucaderi**
Hi all, I have implemented serialisation/deserialisation in the roaring bitmap. Please review the code and hopefully merge it. Thank you.

### CODE REVIEW

**Comment by lemire** [comment]
@lucaderi  Great. If you do not get around to satisfying @fsaintjacques's proposals, I'll have time next week to help with the PR.

**Comment by fsaintjacques** [comment]
This might involve more work than planned. Deserialize functions should also support pointer to the container struct (and possibly buffer) if we want to pre allocate containers.

Ideally the deserialization code should be much more aggressive, e.g. validate that `cardinality < capacity` and such.

**Comment by lemire** [comment]
Let us work with @lucaderi  to make it happen!

**Comment by fsaintjacques** [comment]
Yep, I suggest we open a branch for this.

**Comment by lemire** [comment]
@fsaintjacques Or work on @lucaderi's fork, sending him PRs.

**Comment by lucaderi** [comment]
@fsaintjacques Can you please check if the code is now acceptable for merging? Feel free to make any change that you believe is appropriate. Thank you.

**Comment by fsaintjacques** [comment]
You will need to rebase with @owenkaser recent change.

**Comment by lucaderi** [comment]
@fsaintjacques You comments have been very valuable: thanks. I should have addressed your concerns. Please double check.

**Comment by lemire** [comment]
@lucaderi @fsaintjacques 

I do recommend that we use deterministic tests. It is a problem when a test passes once and then fails another time. A good option is to run the test many times,  with many inputs. When tests are cheap enough, loop hundreds or thousands of times with different inputs can help build confidence, without any need for a random value.

**Comment by owenkaser** [comment]
It's probably obvious, but simply seeding a random number generator
deterministically is often a good way to get determinism with the benefits
of random-looking data.

**Comment by lucaderi** [comment]
Your comments your have been addressed.

**Comment by lucaderi** [comment]
I have added ability to iterate elements as always converting to an array and then iterating is not a good idea IMHO.

**Comment by lemire** [comment]
@lucaderi 

> I have added ability to iterate elements as always converting to an array and then iterating is not a good idea IMHO.

No disagreement there. The ability to iterate through the set values was simply not implemented.

**Comment by lucaderi** [comment]
I am not sure if this is the right place to discuss these things (if not let's move offline). The serialisation I have implemented is to dump on disk the bitmap you have in memory. However running some tests I have noticed that in some cases this is not a good idea as the serialised output is too large. Example consider this bitmap

[serialize_len=1904][cardinality=112]
{643,8253,2404017,2411627,4807391,4815001,7210765,7218375,9614139,9621749,12017513,12025123,14420887,14428497,16824261,16831871,148840237,148847847,151243611,151251221,153646985,153654595,156050359,156057969,158453733,158461343,160857107,16086471
7,163260481,163268091,165663855,165671465,297679831,297687441,300083205,300090815,302486579,302494189,304889953,304897563,307293327,307300937,309696701,309704311,312100075,312107685,314503449,314511059,446519425,446527035,448922799,448930409,4513
26173,451333783,453729547,453737157,456132921,456140531,458536295,458543905,460939669,460947279,463343043,463350653,595359019,595366629,597762393,597770003,600165767,600173377,602569141,602576751,604972515,604980125,607375889,607383499,609779263,
609786873,612182637,612190247,744198613,744206223,746601987,746609597,749005361,749012971,751408735,751416345,753812109,753819719,756215483,756223093,758618857,758626467,761022231,761029841,893038207,893045817,895441581,895449191,897844955,897852
565,900248329,900255939,902651703,902659313,905055077,905062687,907458451,907466061,909861825,909869435}

This is a bitmap with 112 elements set, and its serialised version takes ~17 bits per bit set. In other cases even more space is necessary. In other cases, depending on the values set (not just the cardinality but also what bits are set) you can have much more efficient encoding or even worse. 

In a nutshell ass I want to say is that if you want to dump on a buffer the view of what you have in memory, what I have done is probably a good starting point. However if I want to efficiently (i.e. save space as much as possible) dump the bitmap, probably we need to think about something extra. What do you think? Thank you for your time.

**Comment by lemire** [comment]
@lucaderi  Could you open a separate issue if you want to discuss compression efficiency?

I think it is best to "stay on topic" otherwise it gets difficult later to figure out what was said, when and why.

**Comment by lucaderi** [comment]
@lemire Sure but the question is: serialization must be space-efficient or it must be a 1:1 mapping of the datastructures in memory?

**Comment by lemire** [comment]
@lucaderi  In Java and Go, we have a 1:1 mapping so that you can effectively skip the deserialization and use memory-file mapping.

I do not expect your serialization to be byte-by-byte compatible with the Java and Go versions,  but that might be something to aim for later. This being said if you used 1:1 mapping, you should be somewhat close to the Java and Go versions... and that's just great.

To compress better, Roaring supports run compression 

https://github.com/RoaringBitmap/CRoaring/blob/master/include/roaring.h#L107-L112

That's not yet working in C Roaring. I think that @owenkaser  has done a lot of hard work, but there is more to implement.

We have done some rather extensive testing, using the Java code, and a simple 1-to-1 serialization with array, bitset and run containers seems to offer satisfactory compression and performance in realistic scenarios. 

Speaking for myself, I am more than a bit open to discuss other storage optimizations, including strategies that are not 1-to-1 memory-vs-disk, but if you have implemented a simple serialization strategy that works, I would work on making sure we can merge it into the project.

Please don't be shy about opening an issue specifically to discuss alternative compression strategies.

**Comment by lemire** [comment]
@lucaderi  BTW I will get to reviewing your code one way or the other... Sorry if I have not done so yet.

**Comment by lemire** [comment]
@fsaintjacques  @lucaderi 

Ok. I have spent an hour or so with this PR and this looks very good to me and I recommend pulling this. There is a lot to like in this code. 

Anyhow, @fsaintjacques, do you still have any major concerns that absolutely must be addressed prior to merging this PR? 

@lucaderi  I'd be tempted to slightly tweak your serialization code after the merge so that it is compatible with Java and Go. This can be done later.

As for efficiency, yes, we could use a "compressed" format. But this can be otherwise implemented as another feature.

**Comment by lemire** [comment]
@fsaintjacques 

Note that I still recommend a merge, and I am still asking if you have a critical objection to merging this.

(Just to be clear with respect to my latest comments.)

**Comment by fsaintjacques** [comment]
Most glaring issues have been addressed, it remains to add fuzzing for this.

**Comment by lemire** [comment]
@lucaderi 

Do you want to join the CRoaring team? Though you'd still be able to submit PR, you could also contribute directly to the project.

Note that you have also been invited to the organisation (https://github.com/orgs/RoaringBitmap/).

Cheers!

**Comment by lucaderi** [comment]
@lemire thank your for the offer. I am not a bitmap/compression expert, just following your work since  a few years and I admire what you did. I think that your roaring bitmap is very interesting and I was planning to use for indexing network packets. I will probably send you some more contributions in the near future,  as my project moves forwards. As of your invitation, how can I join your organisation (I don't see an invitation on my github account)? Many thanks to all of you.

**Comment by lemire** [comment]
@lucaderi 

Great. Please go to https://github.com/RoaringBitmap It should self-evident then. If not, please get in touch with me at lemire@gmail.com, I'll work on resolving the issue. Also, check your spam folder.

**Comment by lucaderi** [comment]
Found. Thank you!

**Comment by lemire** [comment]
@lucaderi Note that if you want your membership to be public, you have to switch it on...

https://github.com/orgs/RoaringBitmap/people/lucaderi

This gives you a nice badge on your GitHub profile page. Otherwise, people can't tell that you are part of RoaringBitmap.
